pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -180,96 +204,157 @@ public void close() throws PulsarClientException {
+        reader.thenCompose(reader -> buildFreshTask(reader, completableFuture).thenAccept(lastMessageIds -> {
+            // After get the response of lastMessageIds, put the future and result into `refreshMap`
+            // and then filter out partitions that has been read to the lastMessageID.
+            synchronized (this) {
+                refreshRequests.put(completableFuture, lastMessageIds);
+                filterReceivedMessages(lastMessageIds);
+                // If there is no new messages, the refresh operation could be completed right now.
+                if (lastMessageIds.isEmpty()) {
+                    refreshRequests.remove(completableFuture);
+                    completableFuture.complete(null);
+                }
+            }
+        }));
+        return completableFuture;
+    }
+
+    @Override
+    public void refresh() throws PulsarClientException {
+        refreshAsync().join();
+    }
+
+    private CompletableFuture<Void> readAllExistingMessages(Reader<T> reader) {
         long startTime = System.nanoTime();
         AtomicLong messagesRead = new AtomicLong();
 
-        CompletableFuture<Reader<T>> future = new CompletableFuture<>();
-        reader.getLastMessageIdsAsync().thenAccept(lastMessageIds -> {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        buildFreshTask(reader, future).thenAccept(maxMessageIds -> {
+            readAllExistingMessages(reader, future, startTime, messagesRead, maxMessageIds);
+        });
+        return future;
+    }
+
+    private CompletableFuture<Map<String, TopicMessageId>> buildFreshTask(Reader<T> reader,
+                                                                            CompletableFuture<Void> future) {
+        return reader.getLastMessageIdsAsync().thenApply(lastMessageIds -> {
             Map<String, TopicMessageId> maxMessageIds = new ConcurrentHashMap<>();
             lastMessageIds.forEach(topicMessageId -> {
                 maxMessageIds.put(topicMessageId.getOwnerTopic(), topicMessageId);
             });
-            readAllExistingMessages(reader, future, startTime, messagesRead, maxMessageIds);
+            return maxMessageIds;
         }).exceptionally(ex -> {
             future.completeExceptionally(ex);
             return null;
         });
-        future.thenAccept(__ -> readTailMessages(reader));
-        return future;
     }
 
-    private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Reader<T>> future, long startTime,
+    private void filterReceivedMessages(Map<String, TopicMessageId> lastMessageIds) {
+        // The `lastMessageIds` and `readPositions` is concurrency-safe data types.
+        lastMessageIds.forEach((partition, lastMessageId) -> {
+            if (readPositions.containsKey(partition) && lastMessageId.compareTo(readPositions.get(partition)) <= 0) {
+                lastMessageIds.remove(partition);
+            }
+        });
```

Comment:
You don't need to call this method in a synchronized block. `readPositions` is a concurrent hash map, so you don't need to call `containsKey` and `get` twice.

```java
        lastMessageIds.forEach((partition, lastMessageId) -> {
            MessageId readPosition = readPositions.get(partition);
            if (readPosition != null && lastMessageId.compareTo(readPosition) <= 0) {
                lastMessageIds.remove(partition);
            }
        });
```

Reply:
And I don't understand why did you use `synchronized (this)` for the logic in `refreshAsync`. `refreshRequests` is a `LinkedHashMap` that is not thread safe, but you didn't synchronize all places that access `refreshRequests` like `readTailMessages`. I think we can simply change it to a concurrent hash map.

Reply:
Good suggestion!